Clarify usage of "insertion steps" in the HTML Standard

sylph01 opened the pull request:
(partly) closes #2771 .

- Introduced the concept "HTML element insertion steps"
- Defined what "insertion steps" mean in terms of the HTML standard

TODO: 

- [x] Check if this approach is correct
- [x] For each element that requires a definition of insertion steps, rewrite the current wording using HTML element insertion steps


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/7712/embedded-content.html" title="Last updated on Mar 28, 2022, 6:36 PM UTC (68cf046)">/embedded-content.html</a>  ( <a href="https://whatpr.org/html/7712/0098f44...68cf046/embedded-content.html" title="Last updated on Mar 28, 2022, 6:36 PM UTC (68cf046)">diff</a> )
<a href="https://whatpr.org/html/7712/form-control-infrastructure.html" title="Last updated on Mar 28, 2022, 6:36 PM UTC (68cf046)">/form-control-infrastructure.html</a>  ( <a href="https://whatpr.org/html/7712/0098f44...68cf046/form-control-infrastructure.html" title="Last updated on Mar 28, 2022, 6:36 PM UTC (68cf046)">diff</a> )
<a href="https://whatpr.org/html/7712/form-elements.html" title="Last updated on Mar 28, 2022, 6:36 PM UTC (68cf046)">/form-elements.html</a>  ( <a href="https://whatpr.org/html/7712/0098f44...68cf046/form-elements.html" title="Last updated on Mar 28, 2022, 6:36 PM UTC (68cf046)">diff</a> )
<a href="https://whatpr.org/html/7712/images.html" title="Last updated on Mar 28, 2022, 6:36 PM UTC (68cf046)">/images.html</a>  ( <a href="https://whatpr.org/html/7712/0098f44...68cf046/images.html" title="Last updated on Mar 28, 2022, 6:36 PM UTC (68cf046)">diff</a> )
<a href="https://whatpr.org/html/7712/infrastructure.html" title="Last updated on Mar 28, 2022, 6:36 PM UTC (68cf046)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/7712/0098f44...68cf046/infrastructure.html" title="Last updated on Mar 28, 2022, 6:36 PM UTC (68cf046)">diff</a> )
<a href="https://whatpr.org/html/7712/media.html" title="Last updated on Mar 28, 2022, 6:36 PM UTC (68cf046)">/media.html</a>  ( <a href="https://whatpr.org/html/7712/0098f44...68cf046/media.html" title="Last updated on Mar 28, 2022, 6:36 PM UTC (68cf046)">diff</a> )
<a href="https://whatpr.org/html/7712/scripting.html" title="Last updated on Mar 28, 2022, 6:36 PM UTC (68cf046)">/scripting.html</a>  ( <a href="https://whatpr.org/html/7712/0098f44...68cf046/scripting.html" title="Last updated on Mar 28, 2022, 6:36 PM UTC (68cf046)">diff</a> )

domenic commented on source:
```diff
@@ -1766,6 +1766,19 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
   data-x="control">controls</span> or U+0020 SPACE).</p>
 
+  <p>An HTML element may have specific <dfn data-x="html element insertion steps">HTML element
+  insertion steps</dfn> defined for the element's local name.</p>
```
We can cross link "local name" by using `<span data-x="concept-element-local-name">local name</span>`.

sylph01 replied:
True, thanks for the advice! Fixed @ a427b96

domenic commented on source:
```diff
@@ -1766,6 +1766,19 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+  <p>An HTML element may have specific <dfn data-x="html element insertion steps">HTML element
+  insertion steps</dfn> defined for the element's local name.</p>
+
+  <p>The <span data-x="concept-node-insert-ext">insertion steps</span> in the HTML Standard, given
```
```suggestion
  <p>The <span data-x="concept-node-insert-ext">insertion steps</span> for the HTML Standard, given
```

(BTW GitHub will not let you accept this change by pressing a button because the file is too large, so you would have to make the change manually.)

sylph01 replied:
Fixed @ 3856209 . I was kind of iffy on which to choose too...

domenic commented on source:
```diff
@@ -1766,6 +1766,19 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+  <ol>
+   <li><p>If <var>insertedNode</var>'s namespace is the <span>HTML namespace</span>, and this
+   standard defines <span data-x="html element insertion steps">HTML element insertion steps</span>
+   for <var>insertedNode</var>'s local name, then run the corresponding <span>HTML element insertion
```
Cross-link local name here too.

sylph01 replied:
Also fixed @ a427b96

domenic reviewed (commented):
This setup looks good!

sylph01 commented:
Now that the setups are done I'm going to fix the definitions in each element next

sylph01 commented:
It seems like we also need to define "HTML element removing steps" and do the same for insertion steps, because section 4.8.12.5, 4.10.7, 4.10.17.3 are written with both "node is inserted" and "node is removed" symetrically. I added the definition @ 53cf966.

As far as I can check from cross-references, all usages of "node is removed" is used symetrically along "node is inserted", so I believe it is fine to fix them altogether in this PR.

----

The ones in 4.10.7, 4.10.17.3 are trivial so I'm going to fix them next. I might need help/opinions on 4.8.4.3.2 (Reacting to DOM mutations, under `img`) and 4.12.1.1 (Processing model, under `script`) because these may involve rewriting other parts around it. I will prepare these questions when I finish the more trivial ones.

sylph01 commented:
Hmm, 4.10.7 (The `select` element) wasn't actually that trivial. This one has a list of steps to run that is shared with the "insertion step", "removing step", and also when the `option` asks for a reset. If we can name the list of things to do when `multiple` element is absent then all of these can point to the same code block, but I'm not sure if we can introduce a name for something that's only used under a certain section.
If it's fine to introduce a name, it will become something like:

> The `select` HTML element insertion steps, given *insertedNode* are:
> 1. If the `select` element's `multiple` attribute is absent, then invoke that `select` element's selectedness setting algorithm.
> ...(same for removing steps and `option` asking for a reset)...
> 
> The `select` element's selectedness setting algorithm are:
> 1. Run the first applicable set of steps from the following list:
> ...(the dl after that)

sylph01 commented:
Rewrote 4.10.17.3 on my machine but GitHub doesn't allow me to push... (seems like there's issues on the server side)

----

I would like comments/suggestions on the following matters:

## [4.8.4.3.2 Reacting to DOM mutations](https://html.spec.whatwg.org/#reacting-to-dom-mutations)

As this section is written as a list of relevant mutations, taking these two bullets that mention insertion/removing then rewriting as insertion steps doesn't work. I suggest rewriting these two bullets as following:

> The element's HTML element insertion steps or HTML element removing steps counts the mutation as a relevant mutation.

then either after the list or under section 4.8.3, add the definition of insertion steps/removing steps:

> The `source` HTML element insertion steps, given *insertedNode* are:
> 1. If insertedNode's next sibling is an img element and their parent is a picture element, then, count this as a relevant mutation for the img element.
> ...

## [4.10.7 The select element](https://html.spec.whatwg.org/#the-select-element)

... is as commented above.

## [4.12.1.1 Processing model](https://html.spec.whatwg.org/#script-processing-model)

My understanding is that events listed here which triggers the user agent preparing the `script` element happen **after** a node (the `script` element itself, or any node or document fragment that is inserted into the `script` element), which means that these events are not a part of the insertion step and does not need a definition. If any insertion step is to be defined, it would be for the node or document fragment that would be inserted into the `script` element. Is this correct / does this make sense?

If so, I believe that the usage of [inserted] here can be left as is, and they will refer to that specific node's insertion steps if need be.

domenic commented on source:
```diff
@@ -1766,6 +1766,35 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+  <span data-x="concept-element-local-name">local name</span>.</p>
+
+  <p>The <span data-x="concept-node-insert-ext">insertion steps</span> for the HTML Standard, given
+  <var>insertedNode</var>, is defined as the following:</p>
```
is defined -> are defined (here and below).

sylph01 replied:
Fixed @ 4a0e579

domenic commented on source:
```diff
@@ -27012,12 +27040,16 @@ interface <dfn interface>HTMLSourceElement</dfn> : <span>HTMLElement</span> {
+   data-x="attr-media-src">src</code> attribute and whose <code
+   data-x="dom-media-networkState">networkState</code> has the value <code
+   data-x="dom-media-NETWORK_EMPTY">NETWORK_EMPTY</code>, then invoke that <span>media element
+   </span>'s <span data-x="concept-media-load-algorithm">resource selection algorithm</span>.</li>
```
No extra whitespace before `</span>`

sylph01 replied:
Oops, also fixed @ 4a0e579

domenic reviewed (commented):
Here are a few nits, will comment on the substantive questions next...

domenic commented:
Very encouraging progress!

---

> If it's fine to introduce a name, it will become something like:

It's definitely fine to introduce definitions that are only used locally :).

The confusing thing about the select section though, is it appears to actually be about option elements, not select elements. I.e., it has the precondition "causing the list of options to gain or lose one or more option elements".

So I think it would actually end up as option HTML element insertion/removal steps, which say something like

> If _insertedNode_'s parent is a select element, or its parent is an optgroup element whose parent is a select element, then run that select element's selectedness setting algorithm.

> If _oldParent_ is a select element, or its parent is an optgroup element whose parent is a select element, then run that select element's selectedness setting algorithm.

---

> As this section is written as a list of relevant mutations, taking these two bullets that mention insertion/removing then rewriting as insertion steps doesn't work. I suggest rewriting these two bullets as following:

I like this a lot.

---

> 4.12.1.1 Processing model

This is trickier, because it isn't necessarily a HTML element that is inserted, and we are so far only making the HTML element insertion rigorous. We don't have a plan for DocumentFragments or text nodes or similar.

I can think of ways to make this rigorous, although they all have some tradeoffs. (E.g., maybe we should add specific steps for handling nodes that are inserted into connected `script` elements, inside "the insertion steps for the HTML Standard"? But then we've spread out the information about when "prepare a script" runs into two separate places, which makes it harder to read...)

IMO we should leave this for later. If it becomes the only remaining reference to `#nodes-are-inserted`, then we can change it to link to DOM's insertion steps, which will still be kind of sloppy spec-writing, but it will be a smaller-scoped bit of sloppy spec, and we can tackle it separately.

sylph01 commented:
Thanks for the suggestions! I added the following:

- 4.10.17.3 Association of controls and forms: 3bda554
- 4.8.4.3.2 Reacting to DOM mutations: e738216
    - I added the definitions of insertion/removing steps under 4.8.3 (the `img` element) so that it aligns with the `source` element's definition.
    - The `source` element has a new insertion step from this change. If this needs a specific ordering I can change them.
- 4.10.7 The select element: 916f7ad

Now I will try to fix the build error I got from the commit that didn't get in yesterday...

sylph01 commented:
> 4.12.1.1 Processing model
> This is trickier, because it isn't necessarily a HTML element that is inserted, and we are so far only making the HTML element insertion rigorous.

True, I was missing the case where the thing inserted is not an HTML element. 

If we are leaving this part as is as of now then I believe this PR is ready for review. One thing I am worried about is in 4.8.4.3.2 (Reacting to DOM mutations) I had to split the definition of "relevant mutations" across multiple sections (the `source` element insertion step and the `img` element insertion step) making it harder to read, but I also think this can't be avoided due to the nature of this definition, and also can be mitigated by using cross-references.

If there's anything else that needs fixing I am ready to fix so please feel free to point them out too. Thanks in advance!

domenic commented on source:
```diff
@@ -1766,6 +1766,35 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+   </p></li>
+  </ol>
+
   <p>A <dfn data-x="nodes are inserted">node <var>A</var> is inserted</dfn> into a node <var>B</var>
```
Can we now delete this paragraph?

I guess we would need to adjust the 4.12.1.1 processing model in the manner I described, by changing it from `<span data-x="nodes are inserted">inserted</span>` to `<span data-x="concept-node-insert-ext">inserted</span>`.

sylph01 replied:
Fixed @ f670528.
There was another usage of `nodes are inserted` under https://html.spec.whatwg.org/#concept-media-load-algorithm (under "Otherwise (mode is children)"), but I think this would also be fine with `concept-node-insert-ext` / `concept-node-remove-ext` because these definitions work the same by pointing to the DOM Standard.
As far as I checked this commit removes all instances of `nodes are inserted` / `nodes are removed` `data-x`s.

domenic commented on source:
```diff
@@ -27012,12 +27040,20 @@ interface <dfn interface>HTMLSourceElement</dfn> : <span>HTMLElement</span> {
+   data-x="dom-media-networkState">networkState</code> has the value <code
+   data-x="dom-media-NETWORK_EMPTY">NETWORK_EMPTY</code>, then invoke that <span>media
+   element</span>'s <span data-x="concept-media-load-algorithm">resource selection
+   algorithm</span>.</li>
```
Nit: newline between the two list items.

sylph01 replied:
Fixed @ 4d87aba

domenic commented on source:
```diff
@@ -27246,6 +27282,24 @@ interface <dfn interface>HTMLImageElement</dfn> : <span>HTMLElement</span> {
+  <ol>
+   <li><p>If <var>insertedNode</var>'s parent is a <code>picture</code> element, then, count this as
+   a <span data-x="relevant mutations">relevant mutation</span> for the <code>img</code>
+   element.</p></li>
```
Let's just say "for _insertedNode_".

sylph01 replied:
Fixed @ 4d87aba

domenic commented on source:
```diff
@@ -28677,11 +28731,9 @@ was an English &lt;a href="/wiki/Music_hall">music hall&lt;/a> singer, ...</code
-   <li><p>The element's parent is a <code>picture</code> element and a <code>source</code> <span
-   data-x="nodes are inserted">element is inserted</span> as a previous sibling.</p></li>
+   <li><p>The element's <span>HTML element insertion steps</span> or <span>HTML element removing
+   steps</span> counts the mutation as a <span data-x="relevant mutations">relevant mutation</span>.
```
```suggestion
   steps</span> count the mutation as a <span data-x="relevant mutations">relevant mutation</span>.
```

sylph01 replied:
Fixed @ 4d87aba

domenic commented on source:
```diff
@@ -28677,11 +28731,9 @@ was an English &lt;a href="/wiki/Music_hall">music hall&lt;/a> singer, ...</code
-   data-x="nodes are inserted">element is inserted</span> as a previous sibling.</p></li>
+   <li><p>The element's <span>HTML element insertion steps</span> or <span>HTML element removing
+   steps</span> counts the mutation as a <span data-x="relevant mutations">relevant mutation</span>.
+   </p></li>
```
No whitespace/newline before closing tags here.

sylph01 replied:
Fixed @ 4d87aba

domenic commented on source:
```diff
@@ -51202,13 +51254,13 @@ interface <dfn interface>HTMLSelectElement</dfn> : <span>HTMLElement</span> {
-  a reset</dfn>, then, if the <code>select</code> element's <code
-  data-x="attr-select-multiple">multiple</code> attribute is absent, the user agent must run the
-  first applicable set of steps from the following list:</p>
+  <p>The <dfn data-x="selectedness setting algorithm">selectedness setting algorithm</dfn>, given a
```
If the data-x text is the same as the definition text content, then we can omit the data-x. See https://github.com/whatwg/wattsi/blob/main/Syntax.md for more documentation.

sylph01 replied:
Thanks for this tip! Fixed @ 4d87aba

domenic commented on source:
```diff
@@ -53651,24 +53726,35 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
-  <p>When a <span>form-associated element</span> or one of its ancestors <span data-x="nodes are
-  removed">is removed</span>, then:
+  <p>The <code>form</code> <span data-x="html element removing steps">HTML element removing
+  steps</span>, given <var>removedNode</var> and <var>oldParent</var>, are:</p>
```
It looks like oldParent is not used, so let's just say "given removedNode, are:"

sylph01 replied:
Fixed @ 4d87aba

sylph01 commented on source:
```diff
@@ -27246,6 +27282,24 @@ interface <dfn interface>HTMLImageElement</dfn> : <span>HTMLElement</span> {
+  <ol>
+    <li><p>If <var>oldParent</var> is a <code>picture</code> element, then, count this as a
+    <span data-x="relevant mutations">relevant mutation</span> for the <code>img</code> element
+    (<var>removedNode</var>).</p></li>
```
Fixed @ 4d87aba

sylph01 commented on source:
```diff
@@ -51202,13 +51254,13 @@ interface <dfn interface>HTMLSelectElement</dfn> : <span>HTMLElement</span> {
+  <ol>
+   <li>If the <code>select</code> element's <code data-x="attr-select-multiple">multiple</code>
+   attribute is absent, run the first applicable set of steps from the following list:</li>
+  </ol>
```
Fixed @ a1292b7. I opted for the minimal change here as there are other issues left, but I might fix it afterwards when other issues are cleaned up.

sylph01 replied:
Re-fixed @ ef691bb, where I rewrote the definition using a single `li` (the second approach).

domenic replied:
This rewrite looks beautiful; thank you.

domenic commented on source:
```diff
@@ -28677,14 +28724,12 @@ was an English &lt;a href="/wiki/Music_hall">music hall&lt;/a> singer, ...</code
 
    <li><p>The element's parent is a <code>picture</code> element and a <code>source</code> element
-   that was a previous sibling is <span data-x="nodes are removed">removed</span>.</p></li>
+   that was a previous sibling is <span data-x="concept-node-remove-ext">removed</span>.</p></li>
```
We should be able to convert this into source HTML element removal steps, similar to what you did for the previous bullet.

domenic commented on source:
```diff
@@ -53651,24 +53714,35 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
-  <p>When a <span>form-associated element</span> or one of its ancestors <span data-x="nodes are
-  inserted">is inserted</span>, then:
+  <p>The <code>form</code> <span data-x="html element insertion steps">HTML element insertion
+  steps</span>, given <var>insertedNode</var>, are:</p>
```
So, this change ends up not being correct. It is designed for when _any_ form-associated element is inserted, not for when a `form` HTML element is inserted.

I think we should move these steps into "the HTML Standard's insertion steps" and "the HTML Standard's removing steps". Then we can add something like

```html
<p class="note">The form owner is also reset by the HTML Standard's <span>insertion steps</span> and <span>removing steps</span>.</p>
```

Right above "When the user agent is to reset the form owner".

sylph01 replied:
Fixed @ 3c41a2d. Some notes:

- I used the `concept-node-insert-ext` / `concept-node-remove-ext` `data-x` here as "the HTML Standard's insertion steps" did not have a `data-x` associated with it. 
- About the ordering: I placed the resetting of the form owner to come after each HTML element's HTML element insertion steps.
    - In the insertion steps, there is a step that would "return", which could be problematic when other steps are added after this happens. If there is a chance that anything could be added after this, we can maybe say "Reset the form owner of the form-associated element, unless the form-associated element's parser inserted flag is set." or something around this line.

domenic replied:
Both of those decisions look good to me!

domenic reviewed (commented):
Sorry for the delay in review. I found a couple more issues!

I also pushed a commit with minor fixes to this branch. Please do `git pull --rebase` before doing any further work and you will get those changes locally.

sylph01 commented:
It looks like checks didn't run on the latest commit. How do I manually run them?

domenic commented:
> It looks like checks didn't run on the latest commit. How do I manually run them?

Sometimes this happens. You can push a small change to make them re-run, or just wait for us to re-run them as part of the review.

domenic commented on source:
```diff
@@ -27012,12 +27052,30 @@ interface <dfn interface>HTMLSourceElement</dfn> : <span>HTMLElement</span> {
+  steps</span>, given <var>removedNode</var> and <var>oldParent</var>, are:</p>
+
+  <ol>
+   <li><p>If <var>removedNode</var>'s next sibling is an <code>img</code> element and
```
I realized its next sibling is no longer an img element. I'm not sure if there's a rigorous way to fix this, but I will just change it to "next sibling _was_ an img element"

domenic reviewed (approved):
LGTM, thanks so much for your patience!
